### ISSUE TITLE
Shadow segfaults when built with clang 13.0 in release mode

### ISSUE THREAD

**Issue by stevenengler** [shadow/shadow#1741]
Shadow (as of at least cec4dc2f956) builds without errors when compiling with clang 13.0 in release mode, but segfaults occasionally (but reproducibly) at runtime. Specifically, it segfaults during any of the lossy tcp tests:

```
The following tests FAILED:
	139 - tcp-blocking-lossy-shadow-ptrace (Failed)
	140 - tcp-blocking-lossy-shadow-preload (Failed)
	146 - tcp-nonblocking-poll-lossy-shadow-ptrace (Failed)
	147 - tcp-nonblocking-poll-lossy-shadow-preload (Failed)
	153 - tcp-nonblocking-epoll-lossy-shadow-ptrace (Failed)
	154 - tcp-nonblocking-epoll-lossy-shadow-preload (Failed)
```

The cause of the segfault appears to be some kind of stack corruption related to the `logger_log()` function. The bug only occurs when compiling with optimizations (not `-O0`) and without the `DEBUG` macro enabled. The segfaults seem to occur anywhere that `logger_log()` is called with floating point arguments, so the `al` register is non-zero when entering `logger_log()` (see the `test` instruction at `logger_log+0`).

### Example

Either run in a Fedora 35 container, or download clang/llvm 13.0 from [here](https://github.com/llvm/llvm-project/releases/tag/llvmorg-13.0.0) and add it to your `PATH`. In `CMakeLists.txt`, change the compile options for release mode from `-O3` to `-O1 -fno-inline-functions` to help with debugging.

```
$ clang -v
$ export CC=clang CXX=clang++
$ ./setup build --test --clean
$ (cd build/src/test/tcp \
   && rm -rf tcp-blocking-lossy-shadow-ptrace.data \
   && gdb --args ../../../../build/src/main/shadow --data-directory=tcp-blocking-lossy-shadow-ptrace.data --interpose-method=ptrace --log-level=debug --use-cpu-pinning false ../../../../src/test/tcp/tcp-blocking-lossy.yaml)

(gdb) b _tcp_logCongestionInfo
Breakpoint 1 at 0x957f04: file /shared/code/shadow/main/src/main/host/descriptor/tcp.c, line 1855.

(gdb) run
Thread 5 "shadow-worker" hit Breakpoint 1, _tcp_logCongestionInfo (tcp=tcp@entry=0x7fffd80280c0)
    at /shared/code/shadow/main/src/main/host/descriptor/tcp.c:1855
1855	    gsize outSize = socket_getOutputBufferSize(&tcp->super);

(gdb) b *logger_log
Breakpoint 2 at 0x96a8a0: file /shared/code/shadow/main/src/lib/logger/logger.c, line 165.

(gdb) b logger_log
Breakpoint 3 at 0x96a8cd: file /shared/code/shadow/main/src/lib/logger/logger.c, line 166.

(gdb) cont
Continuing.

Thread 5 "shadow-worker" hit Breakpoint 2, logger_log (logger=logger@entry=0xb88be0, level=level@entry=LOGLEVEL_DEBUG,
    fileName=0xa0d664 "/shared/code/shadow/main/src/main/host/descriptor/tcp.c",
    functionName=0xa0d810 "_tcp_logCongestionInfo", lineNumber=lineNumber@entry=1867,
    format=0xa0d827 "[CONG-AVOID] cwnd=%d ssthresh=%d rtt=%d sndbufsize=%lu sndbuflen=%lu rcvbufsize=%lu rcbuflen=%lu retrans=%lu ploss=%f fd=%i") at /shared/code/shadow/main/src/lib/logger/logger.c:165
165	                const gchar* format, ...) {

(gdb) info frame 2
Stack frame at 0x7fffed15a920:
 rip = 0x957d7d in _tcp_ackProcessing (/shared/code/shadow/main/src/main/host/descriptor/tcp.c:1792);
    saved rip = 0x958d93
 called by frame at 0x7fffed15a970, caller of frame at 0x7fffed15a8c0
 source language c.
 Arglist at 0x7fffed15a910, args: tcp=tcp@entry=0x7fffd80280c0, host=host@entry=0xba0800, packet=<optimized out>,
    header=header@entry=0x7fffe404dc40
 Locals at 0x7fffed15a910, Previous frame's sp is 0x7fffed15a920
 Saved registers:
  rbx at 0x7fffed15a8e8, rbp at 0x7fffed15a910, r12 at 0x7fffed15a8f0, r13 at 0x7fffed15a8f8, r14 at 0x7fffed15a900,
  r15 at 0x7fffed15a908, rip at 0x7fffed15a918

(gdb) cont
Continuing.

Thread 5 "shadow-worker" hit Breakpoint 3, logger_log (logger=0xb88be0, level=LOGLEVEL_DEBUG,
    fileName=0xa0d664 "/shared/code/shadow/main/src/main/host/descriptor/tcp.c",
    functionName=0xa0d810 "_tcp_logCongestionInfo", lineNumber=1867,
    format=0xa0d827 "[CONG-AVOID] cwnd=%d ssthresh=%d rtt=%d sndbufsize=%lu sndbuflen=%lu rcvbufsize=%lu rcbuflen=%lu retrans=%lu ploss=%f fd=%i") at /shared/code/shadow/main/src/lib/logger/logger.c:166
166	    if (!logger) {

(gdb) info frame 2
No frame at level 2.

(gdb) cont
Continuing.

Thread 5 "shadow-worker" received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
```

**Comment by stevenengler** [comment]
We have found a sway to reproduce this bug: https://godbolt.org/z/x5bPcb5ao

Code:

```c
#include <stdio.h>
#include <stdarg.h>
#include <stdbool.h>

void logger_log(bool skip, int lvl, char* filename, char* fnname, int ln, char* fmt, ...) {
    if (skip) {
        return;
    }

    va_list vargs;
    va_start(vargs, fmt);
    vprintf(fmt, vargs);
    va_end(vargs);
}

int main() {
    double x = 4.0;
    logger_log(false, 1, "A", "B", 42, "Message: %f", x);
    return 0;
}
```

Compile with clang 13.0 and `-O1 -fno-omit-frame-pointer`.

```asm
logger_log:                             # @logger_log
        testb   %al, %al
        jne     .LBB0_3
        testl   %edi, %edi
        je      .LBB0_1
.LBB0_2:
        retq
.LBB0_3:
        movaps  %xmm0, -160(%rbp)
        movaps  %xmm1, -144(%rbp)
        movaps  %xmm2, -128(%rbp)
        movaps  %xmm3, -112(%rbp)
        movaps  %xmm4, -96(%rbp)
        movaps  %xmm5, -80(%rbp)
        movaps  %xmm6, -64(%rbp)
        movaps  %xmm7, -48(%rbp)
        testl   %edi, %edi
        jne     .LBB0_2
.LBB0_1:
        pushq   %rbp
        movq    %rsp, %rbp
        subq    $208, %rsp
        leaq    -208(%rbp), %rax
        movq    %rax, -16(%rbp)
        leaq    16(%rbp), %rax
        movq    %rax, -24(%rbp)
        movabsq $206158430256, %rax             # imm = 0x3000000030
        movq    %rax, -32(%rbp)
        movq    stdout(%rip), %rdi
        leaq    -32(%rbp), %rdx
        movq    %r9, %rsi
        callq   vfprintf
        addq    $208, %rsp
        popq    %rbp
        retq
.LCPI1_0:
        .quad   0x4010000000000000              # double 4
main:                                   # @main
        pushq   %rbp
        movq    %rsp, %rbp
        movsd   .LCPI1_0(%rip), %xmm0           # xmm0 = mem[0],zero
        movl    $.L.str.2, %r9d
        xorl    %edi, %edi
        movb    $1, %al
        callq   logger_log
        xorl    %eax, %eax
        popq    %rbp
        retq
.L.str.2:
        .asciz  "Message: %f"
```

We would have expected this to print the value 4.0, not 0.0.

```
ASM generation compiler returned: 0
Execution build compiler returned: 0
Program returned: 0
Message: 0.000000
```

**Comment by stevenengler** [comment]
This ended up being a miscompilation bug in LLVM. It was not fixed in LLVM 13.0.1, but should be fixed in LLVM 14.0.

https://bugs.chromium.org/p/llvm/issues/detail?id=17
